### ISSUE TITLE
Split descriptions into "Terminology" and "Definition"

### ISSUE THREAD

**Issue by szymczdm** [JacquesCarette/literate-scientific-software#94]
This is somewhat related to #93 in that fields in our chunks have been used without any sort of consistency through the examples.

"Name" of a chunk should be the internal (Drasil) name and should not necessarily be used within the documents. The description ("descr") of a chunk should be broken down into a hierarchy consisting of Terminology and Definition.

Terminology is essentially a short form for referring to a concept; it is used to refer to a concept after it has been defined (although, you can use terminology *without* defining it first).

Definitions should only be used after terminology has been created, as we can't define a term that isn't there. This is where the hierarchy comes into play (Term without Def'n is okay, but Def'n without Term is meaningless).

We may want to change this later on, so that all terminology *must* have a definition, but for now it makes sense that common terminology wouldn't need definitions as they would be assumed to be known.

After this has been done, the examples will need to be cleaned up a bit (first pass can leave "descr" as a synonym for "term" or "definition" depending on what is being used in the examples currently so as not to break everything).

If I've missed anything, or not explained in enough detail, please let me know.

**Comment by JacquesCarette** [comment]
This is clear enough for me (but we talked about it).  The real test will be if Spencer understands it too!

**Comment by smiths** [comment]
Yes, this makes sense to me too.  I especially like that you require terminology, but not a description.  In a document that I worked on with Ned, I had a full data definition of density.  He did not react positively, saying that everyone knows what density is.  I still don't think there is anything wrong with being complete, but there is a point where it just isn't worth the effort to have a full description for every symbol.  Having density defined with just terminology seems like a reasonable short cut.  (Ideally, I think the knowledge base should still include a full description of density with the user only choosing a recipe that actually shows this information in cases where it makes sense for the audience.)

**Comment by szymczdm** [comment]
@smiths I agree with your points on completeness, I just mentioned the lack of "definition" for the time being since we don't necessarily have all of that information at hand and finding/updating it would be a bit of a time sink. Definitely something to do in the future.

@JacquesCarette How would you recommend me going about this change. It seems like no matter how I implement it, I'm going to end up modifying a whole mess of things. I was thinking changing the way ConceptChunk is implemented which (if my intuition is right) would probably be the easiest, since the majority of CCs should be created through a smart constructor (makeCC or nCC), which means updating them shouldn't break as many things. 

However, I'm pretty sure Unitals will all break, and possibly quantities.

On the other hand, I could introduce a new datatype (ie. "VerboseConceptChunk" or something of that ilk), but that would only be a stopgap since eventually we would want to do away with the current ConceptChunks in favour of the new ones.

Pretty sure I just answered my own question, but I would still appreciate any input.

**Comment by JacquesCarette** [comment]
Yes, I think you've answered your own question.

**Comment by szymczdm** [comment]
I've just pushed to a new branch because I broke everything. I took a crack at splitting things and it didn't work, so I started over and got stuck. I haven't figured out a design that will work yet (last attempt and this one have revealed assorted tiny problems throughout).

@JacquesCarette would you mind taking a look and letting me know if you see the fundamental flaw in what I'm trying to do?

**Comment by JacquesCarette** [comment]
Part of the flaw starts at

    class Terminology t where
      getTerm :: t -> Sentence

followed by

    -- BEGIN CONCEPT --
    -- a concept has a description
    class Chunk c => Concept c where
      descr :: (Terminology t) => Simple Lens c t
    -- END CONCEPT --

We are re-defining what a Concept is: it has a 'terminology' and an optional definition.  So you should keep Concept as

    class Chunk c => Concept c where
       descr :: Simple Lens c Sentence

but change the instances.  You might want then to add

    class Concept c => ConceptDefinition c where
      cdefn :: Simple Lens c Sentence

To do that, you'll probably want to split Term into Term and DefinedTerm (as two separate data types), and only make DefinedTerm have an instance of ConceptDefinition.  Then use MUChunk-like ideas for joining these.

**Comment by szymczdm** [comment]
Thanks for the input. I worked myself in circles there for a while even after your explanation (mainly, I misunderstood a few things). I was essentially trying to overload descr as both cdefn and descr on simple terms which was obviously wrong, but it took me a while to figure out why.

I think I've got the infrastructure in place (for the most part), although updating the examples may reveal some more issues I haven't thought of yet. Namely, I can feel something is going to happen when we get blank cdefns from concepts which will need to be handled appropriately. Though that problem could be solved fairly easily depending on how deep it ends up being.

If that sounded like rambling. It was. At this point I'm just getting some thoughts down that were swirling in my head as I was working.

As of now, I'm going to continue updating the examples (specifically starting with GamePhysics and the Data.Drasil libraries) to work in the new changes and see what happens.
